forem/forem
# improve keyboard accessibility of modals 12427 10610

## What type of PR is this? (check all applicable)


## Description

This PR has two major strands:

1. **Updates non-Crayons modals to use the Preact Crayons Modal**. This specifically relates to the listings modal, the admin modals (for profile fields and navigation links), and the "Log in to continue" modal that's presented in many cases where a user isn't signed in (e.g. on bookmark or 'reply' click)
2. **Rolls out focus-trap functionality to all Preact Crayons modals** including those noted above.

It also adds an open/close toggle state to the Storybook Preact Modal component, along with accessibility notes.

**What this PR doesn't cover**:

- The `focus-trap` library currently doesn't work within an iframe. This means that the focus trap hasn't been added to the Mod 'flag user' modal, or the Storybook HTML variant - I'll raise a separate ticket to cover this shortly
- Modals triggered by liquid tags (relevant to the user subscribe and poll liquid tags). These will require some further thought/work - I'll raise a separate ticket to cover this too
- I haven't included the delete modal triggered from the connect chat UI, since it isn't reachable by keyboard yet and should be covered in the broader accessibility refactors for that section

## Related Tickets & Documents

- Closes #12427
- Closes #12794 
- Closes #12771
- Covers most of #10610 (rate limit modals to follow later)

## QA Instructions, Screenshots, Recordings

In each of the modal cases above, the following should happen:

- When a button is clicked that triggers a modal to open, focus is immediately moved to the first interactive element in the modal
- While the modal is open, pressing the `Tab` key or `Shift + Tab` cycles through the interactive elements in the modal and ignores any in the background
- When the modal is open and the user presses the `Escape` key, the modal closes and focus is sent back to the activating button
- When the modal is open and the user clicks a button that closes the modal, focus is sent back to the activating button
- In the case of the listings modal, clicking outside of the modal area should close it (maintaining previous behaviour)

### UI accessibility concerns?

This PR introduces accessible focus management for the modals noted above, test steps should be checked with keyboard navigation alone as well as with mouse usage.

## Added tests?

- [X] Yes - tests added for the Preact modal, and Cypress tests added for the "log in to continue" modal
      have not been included_

## Added to documentation?

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [X] Storybook component docs

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![mouse trap cage descends](https://media1.giphy.com/media/ihGBGmtTnc3xmNX9np/giphy.gif?cid=ecf05e47uer8nhuk2a35t1mn7ihyjr2rfb2gwva62iw7ozee&rid=giphy.gif)
